### PR TITLE
feature: add mtls client cert support

### PR DESCRIPTION
**Pull request by ADD-SP**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-websocket project.

### CODE REVIEW

File: lib/resty/websocket/client.lua [hunk 1]
```diff
@@ -122,11 +123,20 @@ function _M.connect(self, uri, opts)
             sock_opts = { pool = pool }
         end
 
-        if opts.ssl_verify then
+        client_cert = opts.client_cert
+        client_priv_key = opts.client_priv_key
+
+        if client_cert then
+            assert(client_priv_key,
```
Comment (dndx) [inline lib/resty/websocket/client.lua:130]: `assert` should be localized at module level.

File: t/cs.t [hunk 2]
```diff
@@ -2056,3 +2089,234 @@ GET /c
 failed to receive: failed to receive the first 2 bytes: closed
 --- error_log
 failed to send close: bad status code
+
+
+=== TEST 29: mutual TLS without client certs
+--- no_check_leak
```
Comment (dndx) [inline t/cs.t:2095]: Is this required?

File: t/cs.t [hunk 3]
```diff
@@ -2056,3 +2089,234 @@ GET /c
...
+            local bytes, err = wb:send_text(data)
+            if not bytes then
+                ngx.log(ngx.ERR, "failed to send the 2nd text: ", err)
+                return ngx.exit(444)
+            end
+        ';
+    }
+--- user_files eval: $::mtls_user_files
+--- udp_listen: 1953
+--- udp_reply eval
```
Comment (dndx) [inline t/cs.t:2165]: Can this be saved as a variable so they the test below can just refer to the function? Like `--- http_config eval: $::HttpConfig`

File: README.markdown [hunk 4]
```diff
@@ -394,6 +394,21 @@ SSL handshake if the `wss://` scheme is used.
 
     Specifies custom headers to be sent in the handshake request. The table is expected to contain strings in the format `{"a-header: a header value", "another-header: another header value"}`.
 
+* `client_cert`
+
+    Specifies a client certificate chain cdata object that will be used while handshaking with remote server.
```
Comment (dndx) [inline README.markdown:399]:
```suggestion
    Specifies a client certificate chain cdata object that will be used while TLS handshaking with remote server. 
```

File: lib/resty/websocket/client.lua [hunk 5]
```diff
@@ -151,7 +162,13 @@ function _M.connect(self, uri, opts)
         if not ssl_support then
             return nil, "ngx_lua 0.9.11+ required for SSL sockets"
         end
-        ok, err = sock:sslhandshake(false, host, ssl_verify)
+        if client_cert then
+            ok, err = sock:setclientcert(client_cert, client_priv_key)
+            if not ok then
+                return nil, "ssl client cert failed: " .. err
```
Comment (dndx) [inline lib/resty/websocket/client.lua:168]:
```suggestion
                return nil, "failed to set TLS client certificate: " .. err
```

**Review by dndx — APPROVED** [review]

File: t/cs.t [hunk 6]
```diff
@@ -2056,3 +2089,130 @@ GET /c
 failed to receive: failed to receive the first 2 bytes: closed
 --- error_log
 failed to send close: bad status code
+
+
+=== TEST 29: mutual TLS with client certs
```
Comment (cppcoffee) [inline t/cs.t:2094]: Need three newline.

File: lib/resty/websocket/client.lua [hunk 7]
```diff
@@ -98,7 +99,8 @@ function _M.connect(self, uri, opts)
         path = "/"
     end
 
-    local ssl_verify, headers, proto_header, origin_header, sock_opts = false
+    local ssl_verify, server_name, headers, proto_header, origin_header, sock_opts = false
```
Comment (cppcoffee) [inline lib/resty/websocket/client.lua:102]: Keep the 80 column limit.

**Comment by dndx** [comment]
@cppcoffee Could you take another look give a ✅ on this?

File: t/cs.t [hunk 8]
```diff
@@ -2056,3 +2065,129 @@ GET /c
...
 --- error_log
 failed to send close: bad status code
+
+
+
+=== TEST 29: mutual TLS with client certs
+--- no_check_leak
+--- http_config eval: $::HttpConfig
+--- config
+    listen 12345 ssl;
```
Comment (zhuizhuhaomeng) [inline t/cs.t:2075]: Please do not use port 12345. This port is blocked by iptables in other lua-nginx-module.

we can use $TEST_NGINX_RAND_PORT_1, please refer to 
https://github.com/openresty/lua-nginx-module/blob/master/t/139-ssl-cert-by.t#L459

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

**Comment by dndx** [comment]
Thank you @zhuizhuhaomeng !

**Pull request merged**